flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherJobTest.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.flink.runtime.jobmaster.utils.TestingJobMasterGateway;
+import org.apache.flink.runtime.jobmaster.utils.TestingJobMasterGatewayBuilder;
+import org.apache.flink.runtime.messages.webmonitor.JobDetails;
+import org.apache.flink.runtime.testtasks.NoOpInvokable;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.core.Is.is;
+
+/**
+ * Test for the {@link DispatcherJob} class.
+ */
+public class DispatcherJobTest extends TestLogger {
+
+	protected final Logger log = LoggerFactory.getLogger(getClass());
+
+	private static final Time TIMEOUT = Time.seconds(10L);
+	private static final JobID TEST_JOB_ID = new JobID();
+
+	@Test
+	public void testInitialSubmissionError() throws ExecutionException, InterruptedException {
+		TestContext testContext = createDispatcherJob();
+		DispatcherJob dispatcherJob = testContext.dispatcherJob;
+
+		Assert.assertThat(dispatcherJob.isRunning(), is(false));
+		Assert.assertThat(dispatcherJob.requestJobStatus(TIMEOUT).get(), is(JobStatus.INITIALIZING));
+		Assert.assertThat(dispatcherJob.requestJobDetails(TIMEOUT).get().getStatus(), is(JobStatus.INITIALIZING));
+		Assert.assertThat(dispatcherJob.getResultFuture().isDone(), is(false));
+
+		// now fail
+		RuntimeException exception = new RuntimeException("Artificial failure in runner initialization");
+		testContext.jobManagerRunnerCompletableFuture.completeExceptionally(exception);
+
+		Assert.assertThat(dispatcherJob.isRunning(), is(false));
+		Assert.assertThat(dispatcherJob.requestJobStatus(TIMEOUT).get(), is(JobStatus.FAILED));
+		Assert.assertThat(dispatcherJob.requestJobDetails(TIMEOUT).get().getStatus(), is(JobStatus.FAILED));
+		Assert.assertThat(dispatcherJob.getResultFuture().isDone(), is(true));
+		ArchivedExecutionGraph aeg = dispatcherJob.getResultFuture().get();
+		Assert.assertThat(aeg.getFailureInfo().getException().deserializeError(ClassLoader.getSystemClassLoader()),	is(exception));
+	}
+
+	@Test
+	public void testCloseWhileInitializing() throws Exception {
+		TestContext testContext = createDispatcherJob();
+		DispatcherJob dispatcherJob = testContext.dispatcherJob;
+
+		Assert.assertThat(dispatcherJob.requestJobStatus(TIMEOUT).get(), is(JobStatus.INITIALIZING));
+
+		CompletableFuture<Void> closeFuture = dispatcherJob.closeAsync();
+		Assert.assertThat(closeFuture.isDone(), is(false));
+
+		// create a jobmanager runner with a mocked JobMaster gateway, that cancels right away.
+		TestingJobManagerRunner jobManagerRunner =
+			new TestingJobManagerRunner(testContext.jobGraph.getJobID(), false);
+
+		// complete JobManager runner future to indicate to the DispatcherJob that the Runner has been initialized
+		testContext.jobManagerRunnerCompletableFuture.complete(jobManagerRunner);
```

Comment:
What happens if we complete this future exceptionally? We should add a test for it.